[Refactoring] Add Codable refactoring action

louisdh opened the pull request:
Inserts the synthesized implementation.
As part of this, fix some ASTPrinter bugs.

rdar://87904700

louisdh commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -1669,7 +1669,7 @@ deriveBodyDecodable_enum_init(AbstractFunctionDecl *initDecl, void *) {
               C, funcDC, C.getKeyedDecodingContainerDecl(),
               caseCodingKeys->getDeclaredInterfaceType(),
-              VarDecl::Introducer::Var, C.Id_nestedContainer);
+              VarDecl::Introducer::Let, C.Id_nestedContainer);
```
The generated code would otherwise produce a warning:
`Variable 'nestedContainer' was never mutated; consider changing to 'let' constant`

The example in the doc comment of this method already said `let nestedContainer`. So this looks like it was a small oversight (which didn't really affect anything so far, I imagine).

louisdh commented on test/refactoring/AddCodableImplementation/Outputs/enum/codable.swift.expected:
```diff
@@ -0,0 +1,65 @@
+      let nestedContainer = try container.nestedContainer(keyedBy: Payload.PlainCodingKeys.self, forKey: Payload.CodingKeys.plain)
+
+      self = Payload.plain(try nestedContainer.decode(String.self, forKey: Payload.PlainCodingKeys._0))
+      case .pair:
```
The indentation is a little weird here, but I don't think that's super important.

bnbarham replied:
Seems like it wouldn't be *too* hard to fix but... 🤷 See above, I wouldn't have bothered at all 😅

louisdh commented on tools/swift-refactor/swift-refactor.cpp:
```diff
@@ -74,6 +74,7 @@ Action(llvm::cl::desc("kind:"), llvm::cl::init(RefactoringKind::None),
                       "replace-bodies-with-fatalError", "Perform trailing closure refactoring"),
            clEnumValN(RefactoringKind::MemberwiseInitLocalRefactoring, "memberwise-init", "Generate member wise initializer"),
            clEnumValN(RefactoringKind::AddEquatableConformance, "add-equatable-conformance", "Add Equatable conformance"),
+           clEnumValN(RefactoringKind::AddExplicitCodableImplementation, "add-explicit-codable-implementation", "Add Explicit Codable Implementation"),
```
Note: this is behaves different from the existing `AddEquatableConformance` action. That is, for Codable we really want the synthesized implementation. This action doesn't make your type conform to `Codable` (which is easy, just make it conform to it manually). It's geared towards "it conforms to Codable, and now I want to insert what the compiler generated".

louisdh commented:
@swift-ci please test

louisdh commented on test/refactoring/AddCodableImplementation/Outputs/basic/codable.swift.expected:
```diff
@@ -0,0 +1,35 @@
+
+    self.firstName = try container.decode(String.self, forKey: User.CodingKeys.firstName)
+    self.lastName = try container.decodeIfPresent(String.self, forKey: User.CodingKeys.lastName)
+    return
```
The `return` at the end is redundant of course. I'm not sure how to best handle that in the ASTPrinter. (we'd have to check if it's the last statement in the scope, I suppose). 

For now, I think it's fine to just keep it. It compiles, but looks a bit weird.

bnbarham replied:
Is it implicit? If it isn't, could be be? And if so, can we skip printing implicit expressions?

louisdh replied:
I ended up adding a (very specific) `SkipImplicitReturn` option to `PrintOptions`. Skipping _all_ implicit `return` stmts made an existing test fail, so I wanted to minimize the potential breakage.

hamishknight replied:
Yeah I'm not sure skipping all implicit returns is correct, as implicit just means "created by the compiler", and we shouldn't skip an implicit return in e.g `if condition { return }`, or one with an expression e.g `return 0`. Would it be possible to narrow this behavior by having the AST printer only drop an implicit return stmt that is the last element of an AbstractFunctionDecl and has no expression?

louisdh replied:
> having the AST printer only drop an implicit return stmt that is the last element of an AbstractFunctionDecl and has no expression?

Ah, good idea. I've done that and removed the option I added (so it will just always skip the printing of such `return`s).

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -3526,6 +3526,197 @@ performChange() {
+      return AddCodableContext(NomDecl);
+    }
+  }
+  // FIXME: support extensions?
+  // (would need to get synthesized nodes from the main decl,
+  // and only if it's in the same file?)
```
In general I think it would be nice to support adding the synthesized code to the extension, I imagine that it would be relatively common.

louisdh replied:
I was hoping to get that support for free, but the synthesized nodes aren't in the extension. It might be trivial to get them from the main decl, but it's just enough trouble that I'd like to keep it as a follow up task.

bnbarham replied:
I assume they're on the main decl, which is trivial to get from the extension (but I haven't checked). Happy to leave as separate, though in that case I'd just leave a `TODO: Add extension support` or similar.

louisdh replied:
Is the `FIXME` I added not enough? Or do you prefer if it really was a `TODO` instead?

bnbarham replied:
I mostly meant remove the `/**/` part. I tend to use "TODO" as "not doing this thing now" and "FIXME" as "fix before merging" but there's no convention there so use whatever you prefer.

bnbarham commented on test/refactoring/AddCodableImplementation/basic.swift:
```diff
@@ -0,0 +1,22 @@
+  let firstName: String
+  let lastName: String?
+}
+// RUN: %refactor -add-explicit-codable-implementation -source-filename %s -pos=3:8 > %t.result/codable.swift
```
Alex added a nice `refactor-check-compiles` that we use for the convert async refactorings. I also prefer `CHECK` to output + diff, but maybe I'm the only one 😆

louisdh replied:
Outputting each run to a separate file makes it very easy to spot bugs (and e.g. to copy the code). Having that all inline in comments would make it harder to understand, I think. 🙂

bnbarham replied:
It's not too bad for these, but for many refactorings *most* of the output doesn't actually matter. I guess this one has a lot compared to the file as a whole though. And personally I find copy pasting often misses bugs 🤷.

Anyway, I'm fine with the separate file, but I would like to check that it compiles.

louisdh replied:
> but I would like to check that it compiles.

As we discussed offline, `refactor-check-compiles` doesn't currently work when outputting to a file. That might be a trivial fix to the test infrastructure, but not something I want to attempt in this PR.

bnbarham reviewed (commented):
Very cool. Thanks for all the work in adding the expression printing too!

There's a few comments but they're all very minor. If you could split the expression printing fixes out into at least a separate commit that would be good, though it was pretty obvious what related to what in this case.

Is the refactoring only valid on the name of the container?

louisdh commented on lib/IDE/Refactoring.cpp:
```diff
@@ -3526,6 +3526,197 @@ performChange() {
+getInsertStartLoc() {
+  SourceLoc MaxLoc = StartLoc;
+  for (auto Mem : Range) {
+    if (Mem->getEndLoc().getOpaquePointerValue() >
```
This was taken from `AddEquatableContext`, and I'd rather just keep this as-is. I don't think it has access to a `SourceManager` here, which would make it slightly more involved to change it as you describe.

hamishknight replied:
You should be able to get it by jumping through a couple of pointers from the `DeclContext`, e.g `DC->getASTContext().SourceMgr`

louisdh commented:
@swift-ci please test

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4275,8 +4282,21 @@ void PrintAST::visitSequenceExpr(SequenceExpr *expr) {
 }
 
+void PrintAST::printBaseExprForTypeAccess(Expr *expr) {
+  if (auto *tExpr = dyn_cast<TypeExpr>(expr)) {
+    if (auto metaType = tExpr->getType()->castTo<AnyMetatypeType>()) {
+      // Don't print `.Type` for an expr.
+      printType(metaType->getInstanceType());
```
I believe `TypeExpr` has a `getInstanceType()` method you could use here. That being said, it kind of seems like this should be the default printing logic for `TypeExpr` anyway (in `visitTypeExpr`), what do you think?

louisdh replied:
Ah, yes, putting this in `visitTypeExpr` is much nicer and works just as well. Thanks!

hamishknight commented on lib/IDE/Refactoring.cpp:
```diff
@@ -3526,6 +3527,178 @@ performChange() {
+  /// Start location of declaration context brace
+  SourceLoc StartLoc;
+
+  /// Array of all conformed protocols
+  SmallVector<swift::ProtocolDecl *, 2> Protocols;
+
+  /// Range of internal members in declaration
+  DeclRange Range;
+
+  bool conformsToCodableProtocol() {
+    for (ProtocolDecl *Protocol : Protocols) {
+      if (Protocol->getKnownProtocolKind() == KnownProtocolKind::Encodable ||
+          Protocol->getKnownProtocolKind() == KnownProtocolKind::Decodable) {
+        return true;
+      }
+    }
+    return false;
+  }
```
Could this use `ModuleDecl::lookupConformance(Type, ...)` instead of searching the conformances manually? That would also let you dig out the `init(from:)` or `encode(to:)` witness and check them for `isSynthesized` if we only want to provide this refactoring in cases where it's a synthesized conformance.

louisdh replied:
Ah, you're totally right. With this current implementation we'll provide the refactor option when `Codable` is entirely manually implemented (so there's nothing synthesized). 

I realize it's not ideal in this state, but I'd like to fix this in a follow up PR.

bnbarham replied:
Looks like it would appear but do nothing in this state. It's a fairly narrow condition anyway (only on a decl conforming to encodable/decodable), so implementing in a follow up sounds reasonable.

louisdh committed:
[ASTPrinter] Fix expression printing for metatype access

And fix operators not printing when we can't get a `MemberOperatorRef`.

louisdh committed:
[Refactoring] Add Codable refactoring action

Inserts the synthesized implementation.
As part of this, fix some ASTPrinter bugs.

rdar://87904700

louisdh committed:
[Refactoring] Don't copy CursorInfo for AddEquatable

louisdh commented on lib/IDE/Refactoring.cpp:
```diff
@@ -3526,6 +3526,197 @@ performChange() {
+}
+
+AddCodableContext AddCodableContext::
+getDeclarationContextFromInfo(ResolvedCursorInfo Info) {
```
Fixed the `AddEquatableContext` in a separate commit.

louisdh committed:
[ASTPrinter] Simplify body check

louisdh commented:
@swift-ci please test

bnbarham commented:
@swift-ci please test linux

hamishknight commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -4151,7 +4154,12 @@ void PrintAST::visitLoadExpr(LoadExpr *expr) {
 
 void PrintAST::visitTypeExpr(TypeExpr *expr) {
-  printType(expr->getType());
+  if (auto metaType = expr->getType()->castTo<AnyMetatypeType>()) {
+    // Don't print `.Type` for an expr.
+    printType(metaType->getInstanceType());
+  } else {
+    printType(expr->getType());
+  }
```
Just noticed this condition won't actually fail as `castTo` will never return a null type. Additionally it will crash if the `TypeExpr`'s type has not been set yet (which I don't *think* can be the case for the Codable synthesis in particular, but in general we should handle it).

We ought to print the instance type if it has been set, or else fallback to printing the `TypeRepr` instead, e.g:

```cxx
  if (auto ty = expr->getInstanceType()) {
    printType(ty);
  } else {
    expr->getTypeRepr()->print(Printer, Options);
  }
```

I'm happy for this to be done in a follow-up though.

bnbarham commented:
@swift-ci please test Linux platform
